crates/sui-cluster-test/src/faucet.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use super::{
+    config::{ClusterTestOpt, Env},
+    helper::verify_gas_coin,
+    wallet_client::WalletClient,
+};
+use anyhow::bail;
+use async_trait::async_trait;
+use clap::*;
+use std::collections::HashMap;
+use std::sync::Arc;
+use sui_faucet::FaucetResponse;
+use sui_types::{base_types::encode_bytes_hex, gas_coin::GasCoin, object::Owner};
+use tokio::time::{sleep, Duration};
+use tracing::{debug, info};
+
+pub struct FaucetFactory;
+
+impl FaucetFactory {
+    pub fn create(
+        options: &ClusterTestOpt,
+        faucet_url: Option<String>,
+    ) -> Arc<dyn Faucet + Sync + Send> {
+        match (&options.env, faucet_url) {
+            (Env::NewLocal, None) => Arc::new(DummyFaucet::new()),
+            (Env::Prod, Some(url)) => Arc::new(RemoteFaucet::new(url)),
+            (Env::Staging, Some(url)) => Arc::new(RemoteFaucet::new(url)),
+            (Env::CustomRemote, Some(url)) => Arc::new(RemoteFaucet::new(url)),
+            _ => panic!("Unallowed combination of parameters."),
+        }
+    }
+}
+
+/// Faucet Abstraction for cluster test
+#[async_trait]
+pub trait Faucet {
+    async fn request_sui_coins(
+        &self,
+        client: &WalletClient,
+        minimum_coins: Option<usize>,
+    ) -> Result<Vec<GasCoin>, anyhow::Error>;
+}
+
+/// Client for a remote faucet that is accessible by
+/// POST requests
+pub struct RemoteFaucet {
+    remote_url: String,
+}
+
+impl RemoteFaucet {
+    fn new(url: String) -> Self {
+        info!("Use remote faucet: {}", url);
+        Self { remote_url: url }
+    }
+}
+
+#[async_trait]
+impl Faucet for RemoteFaucet {
+    /// Request test SUI coins from facuet.
+    /// It also verifies the effects are observed by gateway/fullnode.
+    async fn request_sui_coins(
+        &self,
+        client: &WalletClient,
+        minimum_coins: Option<usize>,
+    ) -> Result<Vec<GasCoin>, anyhow::Error> {
+        let gas_url = format!("{}/gas", self.remote_url);
+        debug!("Getting coin from remote faucet {}", gas_url);
+        let address = client.get_wallet_address();
+        let data = HashMap::from([("recipient", encode_bytes_hex(&address))]);
+        let map = HashMap::from([("FixedAmountRequest", data)]);
+
+        let response = reqwest::Client::new()
+            .post(&gas_url)
+            .json(&map)
+            .send()
+            .await
+            .unwrap()
+            .json::<FaucetResponse>()
+            .await
+            .unwrap();
+
+        if let Some(error) = response.error {
+            panic!("Failed to get gas tokens with error: {}", error)
+        }
+
+        // Let fullnode sync
+        sleep(Duration::from_secs(2)).await;
```

Comment:
we shouldn't sleep, its unreliable. we should probably poll instead.

Reply:
The future improvement plan is to subscribe to fullnode for those txns